<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Surrogate Assay for Genotoxicity Assessment Package for Array Transformation Predictions • saga</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Surrogate Assay for Genotoxicity Assessment Package for Array Transformation Predictions">
<meta property="og:description" content='The SAGA package covers a line of functions for automated severity assessment and transformation
    predictions of surrogate assays. The user may use the functions of this package to upload arrays in order
    to classify them either as "transforming or nontransforming". The package relies on a well-balanced and optimized
    model of surrogate assays and thus serves as a robust basis for the prediction of un-trained data.'>
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">saga</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/saga_vignette.html">SAGA Vignette</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mytalbot/saga_package/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="saga" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#saga" class="anchor"></a>SAGA <img src="https://talbotsr.com/saga_package/logo.png" align="right" height="139">
</h1></div>
<div id="surrogate-assay-for-genotoxicity-assessment" class="section level3">
<h3 class="hasAnchor">
<a href="#surrogate-assay-for-genotoxicity-assessment" class="anchor"></a>Surrogate Assay for Genotoxicity Assessment</h3>
<p>In our paper <strong>“Predicting genotoxicity of integrating viral vectors for stem cell gene therapy using gene expression-based machine learning”</strong> we report an improved in vitro test to determine the risk of insertional mutagenesis of integrating vectors for gene therapy. SAGA builds on the well accepted cell culture protocol of the in vitro immortalization assay (IVIM) but screens for the deregulation of oncogenic gene expression signatures. We demonstrate a new bioinformatic approach to correctly classify the mutagenic potential of retroviral vectors used in previous and current clinical trials.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>The package requires <strong>R&gt;=3.6</strong>.</p>
<p>You can download the development version from <a href="https://github.com/mytalbot/saga_package/">GitHub</a> or use:</p>
<div class="sourceCode" id="cb1"><pre class="r"><span class="co"># install.packages("devtools")</span>
<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"mytalbot/saga_package"</span>, <span class="kw">build_vignettes</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">saga</span>)</pre></div>
<p>If you want the Vignette installed in R as well, set the build_vignettes = TRUE (access browseVignettes(“saga”)). Otherwise, the Vignette will only be accessible via the <a href="https://talbotsr.com/saga_package/index.html">website</a>.</p>
<p>If you don’t have or want to use devtools, SAGA can be installed using the source file (i.e. in RStudio).<br><a href="https://github.com/mytalbot/saga_package/tree/master/sourcefiles">It can be downloaded here.</a></p>
<div id="important-note" class="section level3">
<h3 class="hasAnchor">
<a href="#important-note" class="anchor"></a>Important Note</h3>
<p>Please note, that the SAGA package requires dependencies. Some functions can mask each other (depending on the local R setup). This may cause <strong>warnings</strong> (not errors). These can be ignored since they do not hamper the function of the SAGA package. Warnings may be caused mainly by the HTSanalyzeR and mnormt dependencies.</p>
<p><strong>Since the mnormt package has not yet been updated to R4.0</strong>, the only solution to work with saga on R4.0 is the installation of a previous version of mnormt. Make sure not to update mnormt when installing saga with devtools!</p>
<div class="sourceCode" id="cb2"><pre class="r"><span class="no">PackageUrl</span> <span class="kw">&lt;-</span> <span class="st">"https://cran.r-project.org/src/contrib/Archive/mnormt/mnormt_1.5-7.tar.gz"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="no">PackageUrl</span>, <span class="kw">repos</span> <span class="kw">=</span> <span class="kw">NULL</span>,<span class="kw">type</span><span class="kw">=</span><span class="st">"source"</span>)</pre></div>
</div>
<div id="saga-vignettes-and-working-examples" class="section level3">
<h3 class="hasAnchor">
<a href="#saga-vignettes-and-working-examples" class="anchor"></a>SAGA Vignettes and working examples</h3>
<p>If the Vignette was built during the installation process, it can be accessed like any other Vignette using:</p>
<div class="sourceCode" id="cb3"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span>(<span class="st">"saga"</span>)</pre></div>
<p>The <a href="https://talbotsr.com/saga_package/articles/saga_vignette.html">Vignette</a> is also available on the <a href="https://talbotsr.com/saga_package/index.html">SAGA website</a>.</p>
<p>Working examples can be found in the README file or in the Vignette.</p>
<p>Sample files can be <a href="https://github.com/mytalbot/saga_package/tree/master/samples">downloaded from the GitHub repository</a>.</p>
</div>
<div id="saga-core-data" class="section level3">
<h3 class="hasAnchor">
<a href="#saga-core-data" class="anchor"></a>SAGA core data</h3>
<p>You’ll also need the SAGA core data. They are too large for the main package. Download them here (also from GitHub):</p>
<div class="sourceCode" id="cb4"><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"mytalbot/sagadata"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sagadata</span>)</pre></div>
</div>
</div>
<div id="example-analysis-array-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#example-analysis-array-prediction" class="anchor"></a>Example Analysis &amp; Array Prediction</h2>
<p>The following working examples demonstrate the general workflow of SAGA analysis.</p>
<p>You can download an example data set of four arrays plus a ready-to-run SampleInformation.txt file from the GitHub repository.</p>
<p><a href="https://github.com/mytalbot/saga_package/tree/master/samples">Download</a> the sample files to a folder on your machine, <strong>unzip them and specify the path</strong> (see below) to the selected folder in the script below. The sample arrays had to be zipped because of the GitHub size limitations for files.</p>
<p>Note: Make sure that the phenoTest package is sourced as library(phenoTest) before running the script. This is required for GESEA analysis.</p>
<div id="using-the-wrapper-function" class="section level3">
<h3 class="hasAnchor">
<a href="#using-the-wrapper-function" class="anchor"></a>Using the wrapper function</h3>
<div class="sourceCode" id="cb5"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">phenoTest</span>) <span class="co"># this is mandatory (for GESEA)!</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">saga</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sagadata</span>)

<span class="co">### Path definition (Where are the sample files and the SIF?)</span>
<span class="no">path</span>           <span class="kw">&lt;-</span> <span class="st">"path to your samples"</span>

<span class="co">### saga_gentargets; automatically generates an empty (!) sample information file</span>
<span class="co"># Modify the columns: Filename, Batch, Group, Vector, TrueLabel</span>
<span class="co"># Note: this function does not fully automate the SIF generation. You'll have to</span>
<span class="co"># adjust the files manually or create a new file for your files.</span>
<span class="co"># targets      &lt;- saga_gentargets(smplpath=path)</span>

<span class="co">################################################################################</span>
<span class="co">### Wrapper function - all in one</span>
<span class="co">### to use: uncomment function &amp; execute</span>
<span class="co">################################################################################</span>
<span class="no">mySAGAres</span>      <span class="kw">&lt;-</span> <span class="fu"><a href="reference/saga_wrapper.html">saga_wrapper</a></span>(<span class="no">samplepath</span>, <span class="kw">showModel</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">doGESEA</span><span class="kw">=</span><span class="fl">1</span>)</pre></div>
</div>
<div id="more-detailed-saga-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#more-detailed-saga-analysis" class="anchor"></a>More detailed SAGA analysis</h3>
<div class="sourceCode" id="cb6"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">phenoTest</span>) <span class="co"># this is mandatory!</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">saga</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sagadata</span>)
<span class="co">### Using the wrapper function</span>
<span class="co">################################################################################</span>
<span class="co">### Or use these single functions</span>
<span class="co">################################################################################</span>
<span class="co">### Path definition (Where are the sample files and the SIF?)</span>
<span class="no">path</span>           <span class="kw">&lt;-</span> <span class="st">"path to your samples"</span>

<span class="co">### saga_import</span>
<span class="no">rawdata</span>        <span class="kw">&lt;-</span> <span class="fu"><a href="reference/saga_import.html">saga_import</a></span>(<span class="kw">smplpath</span><span class="kw">=</span><span class="no">path</span>, <span class="kw">showjoint</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">SAGA_RAW</span>       <span class="kw">&lt;-</span> <span class="no">rawdata</span>$<span class="no">SAGA_RAW</span>
<span class="no">TEST_RAW</span>       <span class="kw">&lt;-</span> <span class="no">rawdata</span>$<span class="no">TEST_RAW</span>
<span class="no">pData.joint</span>    <span class="kw">&lt;-</span> <span class="no">rawdata</span>$<span class="no">pData.joint</span>
<span class="no">pData.Test</span>     <span class="kw">&lt;-</span> <span class="no">rawdata</span>$<span class="no">pData.Test</span>

<span class="co">### normalize saga data (plotnumber=1 for normalized boxplot, 2 for tSNE plot)</span>
<span class="no">normalized</span>     <span class="kw">&lt;-</span> <span class="fu"><a href="reference/saga_norm.html">saga_norm</a></span>(<span class="no">SAGA_RAW</span>, <span class="no">TEST_RAW</span>, <span class="no">pData.joint</span>, <span class="kw">plotnumber</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">qunorm.SAGA</span>    <span class="kw">&lt;-</span> <span class="no">normalized</span>$<span class="no">qunorm.SAGA</span>
<span class="no">matrix.SAGA.qn</span> <span class="kw">&lt;-</span> <span class="no">normalized</span>$<span class="no">matrix.SAGA.qn</span>
<span class="no">matrix.test.qn</span> <span class="kw">&lt;-</span> <span class="no">normalized</span>$<span class="no">matrix.test.qn</span>

<span class="co">### remove batch effects</span>
<span class="co"># plotnumber = 1 (tSNE of first 2 dimensions)</span>
<span class="co"># plotnumber = 2 (PCA of first 2 dimensions)</span>
<span class="no">batchnorm</span>      <span class="kw">&lt;-</span> <span class="fu"><a href="reference/saga_batch.html">saga_batch</a></span>(<span class="no">matrix.SAGA.qn</span>, <span class="no">matrix.test.qn</span>, <span class="no">rawdata</span>, <span class="no">pData.joint</span>, <span class="kw">plotnumber</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">index</span>          <span class="kw">&lt;-</span> <span class="no">batchnorm</span>$<span class="no">index</span>
<span class="no">matrix.SAGA</span>    <span class="kw">&lt;-</span> <span class="no">batchnorm</span>$<span class="no">matrix.SAGA</span>
<span class="no">matrix.test</span>    <span class="kw">&lt;-</span> <span class="no">batchnorm</span>$<span class="no">matrix.test</span>

<span class="co">### Sampling and model data collection</span>
<span class="no">model</span>          <span class="kw">&lt;-</span> <span class="fu"><a href="reference/saga_sampling.html">saga_sampling</a></span>(<span class="no">matrix.SAGA</span>, <span class="no">matrix.test</span>)
<span class="no">matrix.train</span>   <span class="kw">&lt;-</span> <span class="no">model</span>$<span class="no">matrix.train</span>
<span class="no">labels.train</span>   <span class="kw">&lt;-</span> <span class="no">model</span>$<span class="no">labels.train</span>
<span class="no">matrix.unknown</span> <span class="kw">&lt;-</span> <span class="no">model</span>$<span class="no">matrix.unknown</span>

<span class="co">### Array predictions with optimized SVM parameters (default settings)</span>
<span class="no">output</span>         <span class="kw">&lt;-</span> <span class="fu"><a href="reference/saga_predict.html">saga_predict</a></span>(<span class="no">samplepath</span>, <span class="no">matrix.train</span>, <span class="no">labels.train</span>, <span class="no">matrix.unknown</span>,
                               <span class="no">pData.Test</span>, <span class="kw">writeFile</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">showRoc</span><span class="kw">=</span><span class="fl">0</span>)
<span class="no">classes</span>        <span class="kw">&lt;-</span> <span class="no">output</span>$<span class="no">predictions</span>
<span class="no">classes</span>

<span class="co">### GESEA</span>
<span class="no">gesea_results</span>  <span class="kw">&lt;-</span> <span class="fu"><a href="reference/saga_gesea.html">saga_gesea</a></span>(<span class="kw">smplpath</span><span class="kw">=</span><span class="no">path</span>, <span class="kw">saveResults</span><span class="kw">=</span><span class="fl">0</span>)</pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/mytalbot/saga_package/">https://​github.com/​mytalbot/​saga_package/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/mytalbot/saga_package/issues">https://​github.com/​mytalbot/​saga_package/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing saga</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Adrian Schwarzer <br><small class="roles"> Author </small>  </li>
<li>Steven Talbot <br><small class="roles"> Maintainer, author </small>  </li>
<li>Axel Schambach <br><small class="roles"> Author, contributor </small>  </li>
<li>Michael Rothe <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Adrian Schwarzer, Steven Talbot, Axel Schambach, Michael Rothe.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
