<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SAGA Vignette • saga</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SAGA Vignette">
<meta property="og:description" content="saga">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">saga</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/saga_vignette.html">SAGA Vignette</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mytalbot/saga_package/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SAGA Vignette</h1>
                        <h4 class="author">Steven R. Talbot</h4>
            
            <h4 class="date">2020-09-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mytalbot/saga_package/blob/master/vignettes/saga_vignette.Rmd"><code>vignettes/saga_vignette.Rmd</code></a></small>
      <div class="hidden name"><code>saga_vignette.Rmd</code></div>

    </div>

    
    
<div id="about-saga-package" class="section level2">
<h2 class="hasAnchor">
<a href="#about-saga-package" class="anchor"></a>About SAGA package</h2>
<p>In our paper <strong>“Predicting genotoxicity of viral vectors for stem cell gene therapy using gene expression-based machine learning”</strong> we report an improved in vitro test to determine the risk of insertional mutagenesis of integrating vectors for gene therapy. SAGA builds on the well-accepted cell culture protocol of the in vitro immortalization assay (IVIM) but screens for the deregulation of oncogenic gene expression signatures. We demonstrate a new bioinformatic approach to correctly classify the mutagenic potential of retroviral vectors used in previous and current clinical trials.</p>
<p>This Vignette covers the basic functions for SAGA analysis in R [1].</p>
</div>
<div id="package-requirements" class="section level1">
<h1 class="hasAnchor">
<a href="#package-requirements" class="anchor"></a>Package Requirements</h1>
<p>To use the SAGA package some Bioconductor dependencies require installation. First, the connection to the Bioconductor database must be established. Then the required packages can be loaded independently. Please note that these packages are only available over Bioconductor and cannot be loaded using CRAN.</p>
<div id="important-note" class="section level3">
<h3 class="hasAnchor">
<a href="#important-note" class="anchor"></a>Important Note</h3>
<p><strong>The SAGA package requires R &gt;=3.6</strong>.</p>
<p>Please note that the SAGA package requires dependencies. Some functions can mask each other (depending on the local R setup). This may cause <strong>warnings</strong> (not errors). These can be ignored since they do not hamper the function of the SAGA package. Warnings may be caused mainly by the HTSanalyzeR and mnormt dependencies.</p>
<p><strong>Since the mnormt package has not yet been updated to R4.0</strong>, the only solution to work with saga on R4.0 is the installation of a previous version of mnormt. Make sure not to update mnormt when installing saga with devtools!</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="no">PackageUrl</span> <span class="kw">&lt;-</span> <span class="st">"https://cran.r-project.org/src/contrib/Archive/mnormt/mnormt_1.5-7.tar.gz"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="no">PackageUrl</span>, <span class="kw">repos</span> <span class="kw">=</span> <span class="kw">NULL</span>,<span class="kw">type</span><span class="kw">=</span><span class="st">"source"</span>)</pre></body></html></div>
</div>
<div id="dependency-installations" class="section level3">
<h3 class="hasAnchor">
<a href="#dependency-installations" class="anchor"></a>Dependency installations</h3>
<p><a href="https://cran.r-project.org/web/packages/BiocManager/vignettes/BiocManager.html">BiocManager</a></p>
<pre><code>install.packages("BiocManager")

BiocManager::install("Biobase")      
BiocManager::install("affy")         
BiocManager::install("affyPLM")      
BiocManager::install("GSEABase")     
BiocManager::install("phenoTest")    
BiocManager::install("BiocGenerics") 
</code></pre>
<p>The <a href="#references">reference section</a> offers more information on the packages Biobase [2], BioGenerics [2], affy [3], affyPLM [4] and phenoTest [5].</p>
<p>The following packages (with dependencies) need to be loaded as well. They can be installed via CRAN. The following links can be used directly in the R console.</p>
<pre><code>install.packages("kernlab")   
install.packages("caret")      
install.packages("bapred")     
install.packages("UsingR")     
install.packages("pROC")       
install.packages("Rtsne")      
install.packages("e1071")       
install.packages("limma")      
install.packages("sva")       
install.packages("gridExtra")  </code></pre>
<p>Again, the <a href="#references">reference section</a> offers more information on the packages limma [6], sva [7], caret [8], phenoTest [9], gridExtra [10], bapred [11], UsingR [12], pROC [13], Rtsne [14-16], GSEABase [17], e1071 [18] and kernlab [19].</p>
</div>
<div id="saga-package-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#saga-package-installation" class="anchor"></a>SAGA Package installation</h2>
<p>SAGA can be downloaded from Github using the following commands. If you want the Vignette installed in R as well, set the build_vignettes = TRUE. Otherwise, the Vignette will only be accessible via the website.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co"># install.packages("devtools")</span>
<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"mytalbot/saga_package"</span>, <span class="kw">build_vignettes</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">saga</span>)</pre></body></html></div>
<p>If you don’t have or want to use devtools, SAGA can be installed using the source file (i.e., in RStudio). <a href="https://github.com/mytalbot/saga_package/tree/master/sourcefiles">It can be downloaded here.</a></p>
<p>Further, the package sagadata (containing the SAGA core data) needs to be loaded separately, as it is too large to fit in the saga package. sagadata can be obtained from GitHub as well:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"mytalbot/sagadata"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sagadata</span>)</pre></body></html></div>
<p>For more information on SAGA, visit the <a href="https://talbotsr.com/saga_package/index.html">website</a>.</p>
</div>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>SAGA</code> analysis starts with raw data obtained from regular gene expression microarray analysis acquired by Agilent platforms. This is important since the functions in this package will only work within the Agilent format. In the current version other platforms were not specifically included into the capacities of this package. If you wish to use other platforms you’ll have to adapt the functions to your specific needs.</p>
<p>As described in the paper, we optimized a training set of arrays to discriminate between potentially dangerous gene therapy vectors and MOCK data to obtain a measurement for vector safety. Vector safety is therefore described as either <em>transforming</em> or <em>untransforming</em> in the package.</p>
<p>The basic SAGA classifier was obtained by various Machine Learning algorithms as well as feature assessment techniques as described in our paper. In essence, a toplist of 9-11 probes was obtained that can predict untrained samples. The <code>SAGA</code> package uses its default data set and the probe toplist to predict user sample data.</p>
<p>Although the arrays provide 39428 probes, the annotations were (<strong>de novo</strong>) remapped to the latest available standard (Gencode.vM18, GRCm38.p6, released July 2018 and Whole Mouse Genome Oligo Microarray 4x44K v2 (ID 026655, released October 2017)), so that only 36226 annotated probes were considered in the final model building process.</p>
<p>Further, the user may specify his/her own positive and negative controls for more confidence in the predicted results.</p>
</div>
<div id="basic-requirements" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-requirements" class="anchor"></a>Basic requirements</h1>
<p>To use the <code>SAGA</code> package, the user has to create a specific folder with sample data. The samples must follow the Agilent design and must be in *.txt format. Sample names should be numeric.</p>
<p>Possible sample names would be: 9910.txt, 9911.txt, 9912.txt, etc.</p>
<p>Further, a <strong>‘SampleInformation.txt’</strong> or targets file is needed. This file consists of a table with certain columns which are later used to merge the user data with the default data. In the shown plots user data will always be colored black for identification purposes.</p>
<p>It is mandatory to name the targets file: <strong>SampleInformation.txt</strong> and it <span style="color:red"><strong>has</strong></span> to be located in the same folder as the samples.</p>
<p>The user may change the contents of this file manually to adjust his/her batch, group, vector, or class settings.</p>
<p><code>SAGA</code> offers a function to generate a blank SampleInformation.txt file automatically. However, this only works if there is no such file in the sample directory. Otherwise, the already provided file will be used. The generated targets file may then be adjusted manually. Make sure not to change the column names. Column names are case sensitive. There must be no other *.txt files in the sample folder, except the samples and the SampleInformation.txt file.</p>
</div>
<div id="data-structure" class="section level1">
<h1 class="hasAnchor">
<a href="#data-structure" class="anchor"></a>Data structure</h1>
<p>Here is a short example of a minimal targets file in txt-format. The first column must contain the array names. You may specify the names by yourself or let your microarray platform choose the labeling. However, it is best that the names do not contain any special characters such as <em>“§%&amp;-”</em> and that they are numeric (and in ascending order).</p>
<p>Make sure that the targets file has the following format.</p>
<table class="table">
<caption>Example of a SampleInformation.txt file.</caption>
<thead><tr class="header">
<th align="left">SAMPLE_ID</th>
<th align="left">Filename</th>
<th align="left">Batch</th>
<th align="left">Group</th>
<th align="left">Vector</th>
<th align="left">TrueLabel</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">X41</td>
<td align="left">41.txt</td>
<td align="left">1</td>
<td align="left">1</td>
<td align="left">A2_MOCK</td>
<td align="left">untransforming</td>
</tr>
<tr class="even">
<td align="left">X42</td>
<td align="left">42.txt</td>
<td align="left">1</td>
<td align="left">2</td>
<td align="left">A9_SIN.AV.EFS</td>
<td align="left">transforming</td>
</tr>
<tr class="odd">
<td align="left">X43</td>
<td align="left">43.txt</td>
<td align="left">1</td>
<td align="left">3</td>
<td align="left">A1_LTR.RV.SF.eGFP</td>
<td align="left">transforming</td>
</tr>
<tr class="even">
<td align="left">X44</td>
<td align="left">44.txt</td>
<td align="left">1</td>
<td align="left">4</td>
<td align="left">A9_SIN.AV.EFS</td>
<td align="left">transforming</td>
</tr>
<tr class="odd">
<td align="left">X45</td>
<td align="left">45.txt</td>
<td align="left">1</td>
<td align="left">5</td>
<td align="left">A10_SIN.AV.SF</td>
<td align="left">transforming</td>
</tr>
<tr class="even">
<td align="left">X46</td>
<td align="left">46.txt</td>
<td align="left">1</td>
<td align="left">6</td>
<td align="left">A9_SIN.AV.EFS</td>
<td align="left">transforming</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
</tr>
</tbody>
</table>
<p>In the table, the first array (41.txt) is a negative MOCK control, labeled as <em>untransforming</em> in the <strong>TrueLabel</strong> column. Using either <strong>untransforming</strong> or <strong>transforming</strong> as class labels, the user may specify his/her own positive or negative controls. For samples with unknown TrueLable insert NA.</p>
<p>We recommend that users include at least 2 to 3 MOCK (negative) and RV.SF (positive) controls in their own models. The batch correction algorithm may need this for proper integration of the provided data into the SAGA data set.</p>
<p><strong>IMPORTANT for GSEA:</strong> For the Gene Set Enrichment Analysis (GSEA) the user must provide at least three samples, of which one must be a MOCK control and labeled as 1 in the Group column. Every MOCK control needs the Group label 1. All other arrays may have other numeric labels.</p>
</div>
<div id="using-the-functions-of-the-saga-package" class="section level1">
<h1 class="hasAnchor">
<a href="#using-the-functions-of-the-saga-package" class="anchor"></a>Using the functions of the SAGA package</h1>
<p>The following section explains the functions of this package and how to use them in SAGA analysis.</p>
<div id="saga_gentargets" class="section level2">
<h2 class="hasAnchor">
<a href="#saga_gentargets" class="anchor"></a>saga_gentargets</h2>
<p>After pooling the samples in a specific folder, a targets file called <code>SampleInformation.txt</code> must be provided for analysis. Without this file SAGA analysis will fail. Since its construction is a bit tedious, we provide a function for the automatic generation of an empty file. However, the user will then have to adjust this file manually to his/her experimental setup (the function cannot know about the individual experimental designs). This includes the fields: Filename, Batch, Group, Vector and TrueLabel.</p>
<p>The two requirements for this function are:</p>
<ul>
<li>there must be at least one sample file in the folder (however, there should really be more (controls!))</li>
<li>there must not be a <code>SampleInformation.txt</code> or any other *.txt file in the sample folder, except the samples</li>
</ul>
<pre><code># This function will generate the empty user targets file for the samples.
saga_gentargets(smplpath)
</code></pre>
</div>
<div id="saga_wrapper" class="section level2">
<h2 class="hasAnchor">
<a href="#saga_wrapper" class="anchor"></a>saga_wrapper</h2>
<p>This function enables the user to perform SAGA analysis in one single step. Only the path to the samples and the SampleInformation.txt file must be provided in the folder. Further, the option <code>doGESEA</code> (default setting is 0) gives control over GSEA analysis. In order to perform GSEA analysis, the <code>SampleInformation.txt</code> file has certain requirements which are shown above. One disadvantage of this wrapper function is that the user will loose some flexibility, e.g. in displaying intermediate results like normalization, etc.</p>
<p>The results of GSEA will be generated as separate PDF and text files in the sample folder.</p>
<pre><code># This function performs SAGA analysis in one step.
mySAGAres &lt;- saga_wrapper(smplpath, showModel=0, doGESEA=0)
</code></pre>
</div>
<div id="saga_import" class="section level2">
<h2 class="hasAnchor">
<a href="#saga_import" class="anchor"></a>saga_import</h2>
<p>The <code>saga_import</code> function imports multiple single microarray data files (*.txt format). The user must specify a sample path in the <code>smplpath</code> object to guide the function to the query data. However, the folder should not contain any other files except the <code>SampleInformation.txt</code> file, which can be generated with the <code>saga_gentargets</code> function.</p>
<p>The argument <code>showjoint</code> can be used for quick data visualization. A boxplot will show the joint data set after appending the user samples to the SAGA data set (default value for <code>showjoint</code> is 0).</p>
<pre><code># This will load the default data.
saga_import(smplpath, showjoint=0)
</code></pre>
<p><strong>Please note: </strong> <code>saga_import</code> relies on Agilent raw data in *.txt format. There is no need to process the readout files prior to SAGA analysis.</p>
</div>
<div id="saga_norm" class="section level2">
<h2 class="hasAnchor">
<a href="#saga_norm" class="anchor"></a>saga_norm</h2>
<p>We use quantile normalization as a preprocessing step prior to analysis. The sample data are always normalized together with the SAGA data set. This means that training and test data are merged during the normalization process and will again be separated later for further analysis. This ensures that data are harmonized, and no artificial normalization bias enters the analysis. The <code>normplot</code> option also gives control over a boxplot showing the normalized values of the joint data set (the default argument is 0).</p>
<p>In case of multiple probes per array, the intensities are averaged to a single value. Agilent readouts usually offer four technical probe replicates and a variety of controls. In the end, 39428 normalized and averaged probes will remain, and internal controls are filtered.</p>
<p>Use the plotnumber argument to show either the normalized data in a boxplot (=1) or a t-SNE (=2) plot of the first two dimensions of the combined data sets.</p>
<pre><code><a href="../reference/saga_norm.html"># This function applies quantile normalization on the whole data set.
saga_norm(SAGA_RAW, TEST_RAW, pData.joint, plotnumber=2)</a></code></pre>
</div>
<div id="saga_batch" class="section level2">
<h2 class="hasAnchor">
<a href="#saga_batch" class="anchor"></a>saga_batch</h2>
<p>Usually microarrays are done in batches. A plethora of external influences may (occasionally) distort otherwise equal arrays beyond recognizability. This often leads to false conclusions when not handled correctly, e.g. wrong clustering results.</p>
<p>To avoid such errors we apply batch correction on the joint data set. The user may define his/her own batches within the <code>SampleInformation.txt</code> file and is required to update the batch information <strong>manually</strong>. The batch number must be sequential and start with a value of 1.</p>
<p>The function requires the rawdata as well as the quantile normalized data from the <code>saga_norm</code> function. Use the plotnumber argument to show either PCA (=1) or t-SNE (=2) plots of the combined data sets.</p>
<pre><code># This function will do the batch correction.
saga_batch(matrix.SAGA.qn, matrix.test.qn, rawdata,  pData.joint, plotnumber=2)
</code></pre>
</div>
<div id="saga_sampling" class="section level2">
<h2 class="hasAnchor">
<a href="#saga_sampling" class="anchor"></a>saga_sampling</h2>
<p>This function will take care of the sample preparation of the joint SAGA data set. It will take the batch corrected data from <code>saga_batch</code> and will use the top 9-11 list of SAGA classifier probes to further filter the whole data set. Then, the joint data set will be separated into a training (SAGA data) and a test set (user samples).</p>
<p>The <code>showSagaModel</code> object gives control over a Principal Components Analysis (PCA) plot showing the first principal components of the model data. Also, the axes show the percentage of explained variance (in %) for the plotted principal components.</p>
<p><span style="color:red"><strong>Red</strong></span> dots are potentially genotoxic (transforming) samples, while the rest codes for other SAGA assays that are untransforming.</p>
<pre><code># This function will do the sampling as a preparation for later classification.
saga_sampling(matrix.SAGA, matrix.test)
</code></pre>
<p>The function needs the batch corrected model and test data (<code>matrix.SAGA</code> and <code>matrix.test</code>).</p>
<p>The function output offers a training matrix, labels for the training matrix as well as a matrix with the unknown sample data. These matrices are required for the classification and prediction step of the <code>saga_predict</code> function.</p>
<p><strong>Please note: </strong> The PCA plot is not directly qualified for classification decisions. Although it will give a qualitative overview on the global “positions” of the sample data within the SAGA data context, spatial neighborhood in PCA is not a reliable classifier. There are better ways of assessing classification success than visual similarity in a PCA plot. However, rather often, ambigous data will lie in the zone between transforming and untransforming arrays.</p>
</div>
<div id="saga_predict" class="section level2">
<h2 class="hasAnchor">
<a href="#saga_predict" class="anchor"></a>saga_predict</h2>
<p>For vector safety prediction, a Support Vector Machine (SVM) with a “radial” kernel is used for building the model. We optimized the cost and gamma factors using the SAGA data and generalize these settings for new and untrained samples.</p>
<p>In a second step, the processed user sample data are predicted using the SVM model. The output shows each sample array along with its SVM probability and a label with <strong>transforming/untransforming</strong> information - whereas transforming indicates potential genotoxicity and untransforming potential safety of the vector. The probability value can be used as a quality indicator on how clear the decision of the SVM was for each sample (the cutoff value is 50%, or &gt;0.5 = transforming).</p>
<pre><code># This function will take care of the model building and prediction steps.
saga_predict(smplpath, matrix.train, labels.train, matrix.unknown, pData.Test, writeFile=0, showRoc=0)
</code></pre>
</div>
<div id="saga_gesea" class="section level2">
<h2 class="hasAnchor">
<a href="#saga_gesea" class="anchor"></a>saga_gesea</h2>
<p>For GSEA, we use a gene set of 20 features to predict the genotoxic potential of a vector. The output of GSEA on the microarray data is a normalized enrichment score (NES). For cells transduced with mutagenic vectors like the positive control LTR.RV.SF (gammaretroviral vector with strong promoter/enhancer elements), we observed NES values above 2. We interpret such a result as a strong enrichment and the risk level of a new test vector as likely genotoxic. Vectors eliciting partial deregulation yielded NES values between 0.9 and 2. We interpret this as a weak enrichment and the risk level of a new test vector as potentially genotoxic. NES values below 0.9 were mainly observed for samples transduced with a safer vector configuration (self-inactivating design and weaker promoter). We describe this as no enrichment of our SAGA core set and the risk level of a new test vector as potentially safe. As cutoffs we determined the following thresholds which will be indicated in the GSEA results output.</p>
<ul>
<li>NES &gt;= 0.9: enrichment (transforming)</li>
<li>NES &lt; 0.9 no enrichment (untransforming)</li>
</ul>
<pre><code><a href="../reference/saga_gesea.html"># This function will perform Gene Set Enrichment Analysis (GSEA).
saga_gesea(smplpath, saveResults=0)</a></code></pre>
</div>
<div id="example-analysis-array-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#example-analysis-array-prediction" class="anchor"></a>Example Analysis &amp; Array Prediction</h2>
<p>The following working examples demonstrate the general workflow of SAGA analysis.</p>
<p>You can download an example data set of three arrays plus a ready-to-run SampleInformation.txt file from the GitHub repository.</p>
<p><a href="https://github.com/mytalbot/saga_package/tree/master/samples">Download</a> the sample files to a folder on your machine, <strong>unzip them and specify the path</strong> (see below) to the selected folder in the script below. The sample arrays had to be zipped because of the GitHub size limitations for files.</p>
<p>Note: Make sure that the phenoTest package is sourced as library(phenoTest) before running the script. This is required for GSEA analysis.</p>
<div id="using-the-wrapper-function" class="section level3">
<h3 class="hasAnchor">
<a href="#using-the-wrapper-function" class="anchor"></a>Using the wrapper function</h3>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">phenoTest</span>) <span class="co"># this is mandatory (for GSEA)!</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">saga</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sagadata</span>)

<span class="co">### Path definition (Where are the sample files and the SIF?)</span>
<span class="no">path</span>           <span class="kw">&lt;-</span> <span class="st">"path to sample files"</span>

<span class="co">### saga_gentargets; automatically generates an empty (!) sample information file</span>
<span class="co"># Modify the columns: Filename, Batch, Group, Vector, TrueLabel</span>
<span class="co"># Note: this function does not fully automate the SIF generation. You'll have to</span>
<span class="co"># adjust the files manually or create a new file for your files.</span>
<span class="co"># targets      &lt;- saga_gentargets(smplpath=path)</span>

<span class="co">################################################################################</span>
<span class="co">### Wrapper function - all in one</span>
<span class="co">################################################################################</span>
<span class="no">mySAGAres</span>      <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/saga_wrapper.html">saga_wrapper</a></span>(<span class="kw">smplpath</span><span class="kw">=</span><span class="no">path</span>, <span class="kw">showModel</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">doGESEA</span><span class="kw">=</span><span class="fl">1</span>)</pre></body></html></div>
</div>
<div id="more-detailed-saga-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#more-detailed-saga-analysis" class="anchor"></a>More detailed SAGA analysis</h3>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">phenoTest</span>) <span class="co"># this is mandatory!</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">saga</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sagadata</span>)
<span class="co">### Use the wrapper function...</span>
<span class="co">################################################################################</span>
<span class="co">### ...or these single functions</span>
<span class="co">################################################################################</span>
<span class="co">### Path definition (Where are the sample files and the SIF?)</span>
<span class="no">path</span>           <span class="kw">&lt;-</span> <span class="st">"path to sample files"</span>

<span class="co">### saga_import</span>
<span class="no">rawdata</span>        <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/saga_import.html">saga_import</a></span>(<span class="kw">smplpath</span><span class="kw">=</span><span class="no">path</span>, <span class="kw">showjoint</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">SAGA_RAW</span>       <span class="kw">&lt;-</span> <span class="no">rawdata</span>$<span class="no">SAGA_RAW</span>
<span class="no">TEST_RAW</span>       <span class="kw">&lt;-</span> <span class="no">rawdata</span>$<span class="no">TEST_RAW</span>
<span class="no">pData.joint</span>    <span class="kw">&lt;-</span> <span class="no">rawdata</span>$<span class="no">pData.joint</span>
<span class="no">pData.Test</span>     <span class="kw">&lt;-</span> <span class="no">rawdata</span>$<span class="no">pData.Test</span>

<span class="co">### normalize saga data (plotnumber=1 for normalized boxplot, 2 for tSNE plot)</span>
<span class="no">normalized</span>     <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/saga_norm.html">saga_norm</a></span>(<span class="no">SAGA_RAW</span>, <span class="no">TEST_RAW</span>, <span class="no">pData.joint</span>, <span class="kw">plotnumber</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">qunorm.SAGA</span>    <span class="kw">&lt;-</span> <span class="no">normalized</span>$<span class="no">qunorm.SAGA</span>
<span class="no">matrix.SAGA.qn</span> <span class="kw">&lt;-</span> <span class="no">normalized</span>$<span class="no">matrix.SAGA.qn</span>
<span class="no">matrix.test.qn</span> <span class="kw">&lt;-</span> <span class="no">normalized</span>$<span class="no">matrix.test.qn</span>

<span class="co">### remove batch effects</span>
<span class="co"># plotnumber = 1 (tSNE of first 2 dimensions)</span>
<span class="co"># plotnumber = 2 (PCA of first 2 dimensions)</span>
<span class="no">batchnorm</span>      <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/saga_batch.html">saga_batch</a></span>(<span class="no">matrix.SAGA.qn</span>, <span class="no">matrix.test.qn</span>, <span class="no">rawdata</span>, <span class="no">pData.joint</span>, <span class="kw">plotnumber</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">index</span>          <span class="kw">&lt;-</span> <span class="no">batchnorm</span>$<span class="no">index</span>
<span class="no">matrix.SAGA</span>    <span class="kw">&lt;-</span> <span class="no">batchnorm</span>$<span class="no">matrix.SAGA</span>
<span class="no">matrix.test</span>    <span class="kw">&lt;-</span> <span class="no">batchnorm</span>$<span class="no">matrix.test</span>

<span class="co">### Sampling and model data collection</span>
<span class="no">model</span>          <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/saga_sampling.html">saga_sampling</a></span>(<span class="no">matrix.SAGA</span>, <span class="no">matrix.test</span>)
<span class="no">matrix.train</span>   <span class="kw">&lt;-</span> <span class="no">model</span>$<span class="no">matrix.train</span>
<span class="no">labels.train</span>   <span class="kw">&lt;-</span> <span class="no">model</span>$<span class="no">labels.train</span>
<span class="no">matrix.unknown</span> <span class="kw">&lt;-</span> <span class="no">model</span>$<span class="no">matrix.unknown</span>

<span class="co">### Array predictions with optimized SVM parameters (default settings)</span>
<span class="no">output</span>         <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/saga_predict.html">saga_predict</a></span>(<span class="no">path</span>, <span class="no">matrix.train</span>, <span class="no">labels.train</span>, <span class="no">matrix.unknown</span>,
                               <span class="no">pData.Test</span>, <span class="kw">writeFile</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">showRoc</span><span class="kw">=</span><span class="fl">0</span>)
<span class="no">classes</span>        <span class="kw">&lt;-</span> <span class="no">output</span>$<span class="no">predictions</span>
<span class="no">classes</span>

<span class="co">### GSEA</span>
<span class="no">gsea_results</span>  <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/saga_gesea.html">saga_gesea</a></span>(<span class="kw">smplpath</span><span class="kw">=</span><span class="no">path</span>, <span class="kw">saveResults</span><span class="kw">=</span><span class="fl">0</span>)</pre></body></html></div>
</div>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>[1] R Core Team (2017). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. URL <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>.</p>
<p>[2] Orchestrating high-throughput genomic analysis with Bioconductor. W. Huber, V.J. Carey, R. Gentleman, …, M. Morgan Nature Methods, 2015:12, 115.</p>
<p>[3] Gautier, L., Cope, L., Bolstad, B. M., and Irizarry, R. A. 2004. affy—analysis of Affymetrix GeneChip data at the probe level. Bioinformatics 20, 3 (Feb. 2004), 307-315.</p>
<p>[4] Bolstad, BM (2004) Low Level Analysis of High-density Oligonucleotide Array Data: Background, Normalization and Summarization. Dissertation. University of California, Berkeley.</p>
<p>[5] Evarist Planet (2018). phenoTest: Tools to test association between gene expression and phenotype in a way that is efficient, structured, fast and scalable. We also provide tools to do GSEA (Gene set enrichment analysis) and copy number variation.. R package version 1.30.0.</p>
<p>[6] Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47.</p>
<p>[7] Jeffrey T. Leek, W. Evan Johnson, Hilary S. Parker, Elana J. Fertig, Andrew E. Jaffe, John D. Storey, Yuqing Zhang and Leonardo Collado Torres (2017). sva: Surrogate Variable Analysis. R package version 3.24.4.</p>
<p>[8] Max Kuhn. Contributions from Jed Wing, Steve Weston, Andre Williams, Chris Keefer, Allan Engelhardt, Tony Cooper, Zachary Mayer, Brenton Kenkel, the R Core Team, Michael Benesty, Reynald Lescarbeau, Andrew Ziem, Luca Scrucca, Yuan Tang, Can Candan and Tyler Hunt. (2018). caret: Classification and Regression Training.</p>
<p>[9] Evarist Planet (2013). phenoTest: Tools to test association between gene expression and phenotype in a way that is efficient, structured, fast and scalable. We also provide tools to do GSEA (Gene set enrichment analysis) and copy number variation.. R package version 1.24.0.</p>
<p>[10] Baptiste Auguie (2017). gridExtra: Miscellaneous Functions for “Grid” Graphics. R package version 2.3.</p>
<p>[11] Hornung, R., Boulesteix, A.-L., Causeur, D. (2016) Combining location-and-scale batch effect adjustment with data cleaning by latent factor adjustment. BMC Bioinformatics 17:27.</p>
<p>[12] John Verzani (2018). UsingR: Data Sets, Etc. for the Text “Using R for Introductory Statistics”, Second Edition. R package version 2.0-6.</p>
<p>[13] Xavier Robin, Natacha Turck, Alexandre Hainard, Natalia Tiberti, Frédérique Lisacek, Jean-Charles Sanchez and Markus Müller (2011). pROC: an open-source package for R and S+ to analyze and compare ROC curves. BMC Bioinformatics, 12, p. 77. DOI: 10.1186/1471-2105-12-77</p>
<p>[14] L.J.P. van der Maaten and G.E. Hinton. Visualizing High-Dimensional Data Using t-SNE. Journal of Machine Learning Research 9(Nov):2579-2605, 2008.</p>
<p>[15] L.J.P. van der Maaten. Accelerating t-SNE using Tree-Based Algorithms. Journal of Machine Learning Research 15(Oct):3221-3245, 2014.</p>
<p>[16] Jesse H. Krijthe (2015). Rtsne: T-Distributed Stochastic Neighbor Embedding using a Barnes-Hut Implementation, URL: <a href="https://github.com/jkrijthe/Rtsne" class="uri">https://github.com/jkrijthe/Rtsne</a>.</p>
<p>[17] Martin Morgan, Seth Falcon and Robert Gentleman (2019). GSEABase: Gene set enrichment data structures and methods. R package version 1.46.0.</p>
<p>[18] David Meyer, Evgenia Dimitriadou, Kurt Hornik, Andreas Weingessel and Friedrich Leisch (2019). e1071: Misc Functions of the Department of Statistics, Probability Theory Group (Formerly: E1071), TU Wien. R package version 1.7-3.</p>
<p>[19] Alexandros Karatzoglou, Alex Smola, Kurt Hornik, Achim Zeileis (2004). kernlab - An S4 Package for Kernel Methods in R. Journal of Statistical Software 11(9), 1-20.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Adrian Schwarzer, Steven Talbot, Axel Schambach, Michael Rothe.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
